chapters/high_level_shader_language_comparison.adoc

```diff
@@ -0,0 +1,972 @@
+----
+layout (location = <attachment-index>) out/in <type> <name>;
+----
+
+Example:
+[source,glsl]
+----
+layout (location = 0) out vec4 outPosition;
+layout (location = 1) out vec4 outNormal;
+layout (location = 2) out vec4 outAlbedo;
+
+void main() {
+    outPosition = ...
+    outNormal = ...
+    outAlbedo = ...
+}
+----
+
+===== HLSL
+
+[source,hlsl]
+----
+<type> <name> : SV_TARGET<attachment-index>;
+----
+
+Example:
+[source,hlsl]
+----
+struct FSOutput
+{
+	float4 Position : SV_TARGET0;
+	float4 Normal : SV_TARGET1;
+	float4 Albedo : SV_TARGET2;
+};
+
+FSOutput main(VSOutput input) {
+    output.Position = ...
+    output.Normal = ...
+    output.Albedo = ...
+    return output;
+}
+----
+
+=== Push constants
+
+==== GLSL
+
+[source,glsl]
+----
+layout (push_constant) uniform <structure-type> { <members> } <name>
+----
+
+Example:
+[source,glsl]
+----
+layout (push_constant) uniform PushConsts {
+	mat4 matrix;
+} pushConsts;
+----
+
+==== HLSL
+
+[source,hlsl]
+----
+[[vk::push_constant]] <structure-type> <name>;
+----
+
+[source,hlsl]
+----
+struct PushConsts {
+    float4x4 matrix;
+};
+[[vk::push_constant]] PushConsts pushConsts;
+----
+
+=== Specialization constants
+
+==== GLSL
+
+[source,glsl]
+----
+layout (constant_id = <specialization-constant-index>) const int <name> = <default-value>;
+----
+
+Example:
+[source,glsl]
+----
+layout (constant_id = 0) const int SPEC_CONST = 0;
+----
+
+==== HLSL
+
+[source,hlsl]
+----
+[[vk::constant_id(<specialization-constant-index>)]] const int <name> = <default-value>;
+----
+
+Example:
+[source,hlsl]
+----
+[[vk::constant_id(0)]] const int SPEC_CONST = 0;
+----
+
+=== Sub passes
+
+==== GLSL
+
+[source,glsl]
+----
+layout (input_attachment_index = <input-attachment-index>, binding = <binding-index>) uniform subpassInput <name>;
+----
+
+Example:
+[source,glsl]
+----
+layout (input_attachment_index = 0, binding = 0) uniform subpassInput input0;
+----
+
+==== HLSL
+
+[source,hlsl]
+----
+[[vk::input_attachment_index(<input-attachment-index>)]][[vk::binding(<binding-index>)]] SubpassInput <name>;
+----
+
+Example:
+[source,hlsl]
+----
+[[vk::input_attachment_index(0)]][[vk::binding(0)]] SubpassInput input0;
+----
+
+=== Texture reads
+
+[NOTE]
+====
+Where GLSL uses global functions to access images, HLSL uses member functions of the texture object.
+====
+
+Example:
+
+GLSL:
+[source,glsl]
+----
+layout (binding = 0, set = 0) uniform sampler2D sampler0;
+
+void main() {
+    vec4 color = texture(sampler0, inUV);
+}
+----
+
+HLSL:
+[source,hlsl]
+----
+Texture2D texture0 : register(t0, space0);
+SamplerState sampler0 : register(s0, space0);
+
+float4 main(VSOutput input) : SV_TARGET {
+    float4 color = texture0.Sample(sampler0, input.UV);
+}
+----
+
+[options="header"]
+|====
+| *GLSL*  | *HLSL*
+| texture | Sample
+| textureGrad | SampleGrad
+| textureLod | SampleLevel
+| textureSize | GetDimensions
+| textureProj | n.a.
+| texelFetch | Load
+| sparseTexelsResidentARB | SampleLevel
```

Comment:
? The texture functions have an optional output argument, which can then by queried with a separate function.